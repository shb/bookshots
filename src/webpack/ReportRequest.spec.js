import RequestReport from "./ReportRequest";

describe("RequestReport", () =>
{
  test("isWaitingForTooLong", done =>{
    const req = new RequestReport()
    req.REQUEST_TIMEOUT = 100
    req.startWaiting()
    expect(req.isWaitingForTooLong()).toBe(false)
    setTimeout(() => {
      expect(req.isWaitingForTooLong()).toBe(true)
      done()
    }, 101)
  })

  test("waitForReport reject/throws after waiting for too long", async () => {
    const req = new RequestReport()
    req.REQUEST_TIMEOUT = 100
    try {
      await req.waitForReport()
    } catch (e) {
      expect(e).toMatchSnapshot()
    }
  })

  test("waitForReport resolves after updating the report", done => {
    const req = new RequestReport()
    req.REQUEST_TIMEOUT = 100
    setImmediate(() => {
      req.updateReport("report!")
    })
    req.waitForReport().then(report => {
      expect(report).toEqual("report!")
      done()
    }).catch(done)
  })

  test("waitForReport only resolves to an updated report once", done => {
    const req = new RequestReport()
    req.REQUEST_TIMEOUT = 100
    setImmediate(() => {
      req.updateReport("report!")
    })
    req.waitForReport().then(report => {
      expect(report).toEqual("report!")
    })
    .then(async () => {
      const start = Date.now()
      try {
        await req.waitForReport()
      } catch (e) {
        expect(e).toMatchSnapshot()
        expect(Date.now() - start > 99).toBe(true)
        done()
      }
    })
    .catch(done)
  })

  test("waitForReport start waiting from scratch after a timeout", async () => {
    const req = new RequestReport()
    req.REQUEST_TIMEOUT = 100
    let start
    try {
      start = Date.now()
      await req.waitForReport()
    } catch (e) {
      expect(Date.now() - start).toBeGreaterThan(99)
    }
    try {
      start = Date.now()
      await req.waitForReport()
    } catch (e) {
      expect(Date.now() - start).toBeGreaterThan(99)
    }
  })
})