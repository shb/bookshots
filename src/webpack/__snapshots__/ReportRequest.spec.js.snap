// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RequestReport waitForReport only resolves to an updated report once 1`] = `[Error: Timeout]`;

exports[`RequestReport waitForReport reject/throws after waiting for too long 1`] = `[Error: Timeout]`;
